### DISCUSSION TITLE
Automatic title generation from file name

### DISCUSSION THREAD

**Discussion by xplosionmind** [11ty/eleventy#2241] [Q&A]
Since I have 200+ pages on my website, I am looking for a way to automatically define the `title` variable as if it was set in the front matter, but without having to specify it in every single file. Through the defaults provided by the data cascade, I pinpointed `fileSlug`, that can be used for the permalink. Nevertheless, how can I find a similar function that would pass the exact value of the file name, minus the extension?

Thank you very much!

**Comment by cfjedimaster** [comment]
How about filePathStem?

https://www.11ty.dev/docs/data-eleventy-supplied/#page-variable

**Reply by xplosionmind** [reply]
Definitely far from being the solution, since:
- it yields the full path to the file (and such path should be removed in some way, but it varies)
- it still gives a slugified output…

**Reply by cfjedimaster** [reply]
I'm confused - I get you don't want the dir, we can fix that, but it's not slugified, it's the filename. Are you seeing something diferent? Ie the file is A and filepathstem returns /pathstuff/B ?

**Comment by cfjedimaster** [comment]
Given a folder named posts, you could make `posts.11tydata.js` and automate it for the directory like so - and per your above comment I returned just the filename:

```js
module.exports = {
  eleventyComputed: {
	title: data => data.page.filePathStem.split('/').pop()
  }
};
```

**Reply by xplosionmind** [reply]
How can I make this work globally?

If I create `.11tydata.js` or `11tydata.js` and I place it in the root folder of the website, the file is not recognized…

**Reply by cfjedimaster** [reply]
I don't think you can. However - in your .eleventy.js file, you are allowed to modify collections. You can loop through collections.all. Let me try that real quick.

**Reply by cfjedimaster** [reply]
Actually no  - that would only work if you make a new collection. I've done that before (take collection A, modify it to B).  Why not use a filter instead? Where you display your title, do {{ title | makeTitle }}, and have makeTitle use the pageFileStem.

**Reply by cfjedimaster** [reply]
So given this:

```
{{ page | makeTitle }}
```

I did this in `.eleventy.js`:

```js
 eleventyConfig.addFilter("makeTitle", 
 function(page) {
 	return page.filePathStem.split('/').pop();
 });
```

**Reply by pdehaan** [reply]
I like to put all my source files into a src/ directory and then you should be able to create a src/src.11tydata.js file which will be global.
Then just make sure you set your input directory in the .eleventy.js config file (at the bottom of the exported function), something like:

```
return {
  dir: {
    input: "src", // default "."
    output: "www", // default "_site"
  }
};
```

Otherwise you could try a global _data/eleventyComputed.js file, but I find that semi-problematic and  only use as a last resort.

**Reply by cfjedimaster** [reply]
That's a cool trick there. I do most of my Eleventy stuff in root - but know that it would probably be better to use a subdirectory. This is a great reason. I may blog an example of this.

**Reply by pdehaan** [reply]
@xplosionmind I took a quick look at your https://github.com/xplosionmind/tommi.space/tree/main/_posts directory and created a quick _posts/_posts.11tydata.js directory data file which tries to guess the `title` from the `page.filePathStem` (per @cfjedimaster's excellent suggestion), but will try and preserve the `title` front matter if you already explicitly set that:

This might be semi-close:

```js
const _ = require("lodash");
const deslugify = (str) => _.startCase(str.replace(/-/g, " "));

module.exports = {
  eleventyComputed: {
    title(data) {
      // return data.title || deslugify(data.page?.fileSlug);

      let hadTitle = false;
      const title = data.title || deslugify(data.page?.fileSlug);
      if (data.title) {
        hadTitle = true;
      }
      console.log(`${data.page.filePathStem} => ${title}${hadTitle ? " (had title)" : ""}`);
      return title;
    }
  }
};
```

**NOTE:** You can delete most of that code above, but it was a bit verbose because I wanted extra logging while debugging.

Here's the rough debug output from my Terminal.

```sh
> tommi.space@1.0.0 build
> eleventy --quiet

/_posts/borderlife => Borderlife
/_posts/Che-palle-instagram => Che Palle Instagram
/_posts/Come-cambiare-il-mondo => Come Cambiare Il Mondo
/_posts/Curiosi-Corsari-Consapevoli => Curiosi Corsari Consapevoli
/_posts/Domodossola => Domodossola
/_posts/Frammenti => Frammenti
/_posts/Giffoni => Giffoni
/_posts/Groovy-Pieces => Groovy Pieces
/_posts/Inciampare => Inciampare
/_posts/Iter-per-mentis-casinum => Iter Per Mentis Casinum
/_posts/Keep-calm-and-stop-hating => Keep Calm And Stop Hating
/_posts/Le-torce-non-servono-nello-spazio => Le Torce Non Servono Nello Spazio
/_posts/Making-this-website => Making This Website
/_posts/Maturandi => Maturandi
/_posts/Mnemosyne => Mnemosyne
/_posts/Non-so-più-leggere => Non So Piu Leggere
/_posts/Non-sono-abbastanza => Non Sono Abbastanza
/_posts/Ode-a-Jarvis => Ode A Jarvis
/_posts/Prima-di-scattare-pensa => Prima Di Scattare Pensa
/_posts/Servire => Servire
/_posts/Sogno-ad-occhi-aperti => Sogno Ad Occhi Aperti
/_posts/Solo => Solo
/_posts/Sono-innamorato => Sono Innamorato
/_posts/squali-consapevoli => Squali Consapevoli
/_posts/Svevoismo => Svevoismo
/_posts/The-Bus-Blogger => The Bus Blogger
/_posts/Tirocinanti-saltellanti => Tirocinanti Saltellanti
/_posts/Ugomo => Ugomo
/_posts/Vivi-la-vita => Vivi La Vita
/_posts/Vorrei-avere-culo => Vorrei Avere Culo
/_posts/Avanguardia => L’avanguardia (had title)
/_posts/Brunori-Sa => Brunori Sa, e io? {e noi?} (had title)
/_posts/Cambiare-Vita => Cambiare Vita (had title)
/_posts/Futuro => La sorprendente Prevedibilità del Futuro (had title)
/_posts/Genesi => De Explosiva mente (o Genesi di xplosionmind) (had title)
/_posts/LArte-del-Procrastinare => L’Arte del Procrastinare (had title)
/_posts/Scaricare-film-come-si-deve => Scaricare film, come si deve (had title)
/_posts/The-power-of-decentralization => The power of decentralization (had title)

[11ty] Wrote 350 files in 0.47 seconds (1.3ms each, v1.0.0)
```

---

**UPDATE:** In fact, reading the [`_.startCase`](https://lodash.com/docs/4.17.15#startCase) documentation again, you might not even need to try deslugifying the path first and you can use `return data.title || _.startCase(data.page?.fileSlug);` directly.

**Reply by xplosionmind** [reply]
Thanks a lot for all of your replies!

I managed to make everything work, thanks to @pdehaan’s code.

Nevertheless, I have an include `_includes/head.html`, that I insert with the Liquid `{% render %}` tag inside my “wrapper” layout. In it, among other stuff, there is this tag:

```html
<title>{{ title }} | Tommi’s Space</title>
```

even if I use `{{ page.title }}` or `{{ page.data.title }}`, in every page the output title is ` | Tommi’s Space`.

Why is it not working properly?

**Reply by pdehaan** [reply]
I haven't used LiquidJS's new [`{% render %}` tag](https://liquidjs.com/tags/render.html) much yet, but are you passing the `title` variable to it?
```liquid
{% render 'head.html', title: title %}
```

I'm not sure if it inherits global/local data automatically (like the old `{% include %}` tag) or if you have to pass all the expected variables into it manually.

From their docs:
> ### Variable Scope
> When a partial template is rendered, the code inside it can’t access its parent’s variables and its variables won’t be accessible by its parent. This encapsulation makes partials easier to understand and maintain.

**Reply by xplosionmind** [reply]
I have to pass the `title`, as well as any mother page value, manually inside each `render`… this is very annoying, buy I managed to make it work.

**Comment by Zxce3** [comment]
hey, try this

`_post/post.11tydata.js`
```js
const path = require('path');
const fs = require('fs');
const files = fs.readdirSync(__dirname);

module.exports = {
    permalink: '{{ page.fileSlug }}/index.html',
    eleventyComputed: {
        title: data => {
            const file = files.find(file => file === data.page.fileSlug + '.md');
            return file ? path.basename(file, '.md').replace(/-/g, ' ') : '';
        }
    }
};

// the explanation is like this
// 1. we get the file name from the directory
// 2. we find the file name that matches the file slug
// 3. we get the file name without the extension
// 4. we replace the hyphen with a space
// 5. we return the result

```
